eduvpn/ui/stats.py

```diff
@@ -0,0 +1,138 @@
+import os
+import logging
+import time
+from datetime import timedelta
+
+from ..nm import get_iface, get_ipv4, get_ipv6, get_timestamp
+
+logger = logging.getLogger(__name__)
+
+LINUX_STATS_FILE = "/sys/class/net/{iface}/statistics/{filename}"
+
+
+class NetworkStats():
+    def __init__(self, interval):
+        self.iface = None
+        self.interval = interval
+        self.addresses = {}
+        self.files = {}
+        self.start_bytes = {}
+        self.timestamp = None
+        self.default_text = "N/A"
+
+    def open_file(self, filename):
+        """
+        Helper function to open a statistics network file
+        """
+        filepath = LINUX_STATS_FILE.format(iface=self.iface, filename=filename)
+        if not os.path.isfile(filepath):
+            logger.info(f"Network Stats: {filepath} is not a file")
+            return None
+        return open(filepath, "r")
+
+    def get_human_readable_bytes(self, total_bytes):
+        """
+        Helper function to calculate the human readable bytes.
+        E.g. B, KB, MB, GB, TB.
+        """
+        hr_bytes = total_bytes
+        suffix = "TB"
+        for suffix in ["B", "KB", "MB", "GB"]:
+            if hr_bytes < 1024.0:
+                break
+            hr_bytes /= 1024.0
+
+        return f"{round(hr_bytes, 2)} {suffix}"
+
+    def get_file_bytes(self, filename, typename):
+        """
+        Helper function to get a statistics file to calculate the total data transfer
+        """
+        # If interface wasn't set
+        if self.iface is None:
+            self.iface = get_iface()
```

Comment:
what happens if the iface changes? Does this code handle that properly?  If not, maybe not set the iface but just always get it

Reply:
I don't think the interface will ever change when the vpn connection is active right? This class is alive as long as the connection is

Reply:
ok if it logically makes sense then it's okay, but in general I think it is best to store as minimal state as possible unless it is inside a busy loop, which this is not.

Reply:
This gets run twice (once for upload and download) every second so I want to use as little resources as possible, this is especially the reason why I keep the files open and only close them after the thread is finished.

Reply:
ah right, thats true. maybe it should be set only once during a state transition as described in the other thread.